File: ruby/lib/blackroad/agents.rb
# frozen_string_literal: true

module Blackroad
  # API for managing agents
  class AgentAPI
    def initialize(client)
      @client = client
    end

    # List agents with optional filters
    #
    # @param type [String] Filter by agent type
    # @param division [String] Filter by division
    # @param level [Integer] Filter by level
    # @param status [String] Filter by status
    # @param limit [Integer] Maximum results
    # @param offset [Integer] Offset for pagination
    # @return [Array<Hash>] List of agents
    def list(type: nil, division: nil, level: nil, status: nil, limit: nil, offset: nil)
      params = {}
      params[:type] = type if type
      params[:division] = division if division
      params[:level] = level if level
      params[:status] = status if status
      params[:limit] = limit if limit
      params[:offset] = offset if offset

      response = @client.get("/agents", params)
      response["agents"] || []
    end

    # Get a specific agent by ID
    #
    # @param agent_id [String] The agent ID
    # @return [Hash] Agent data
    def get(agent_id)
      @client.get("/agents/#{agent_id}")
    end

    # Register a new agent
    #
    # @param name [String] Agent name
    # @param type [String] Agent type (default: "ai")
    # @param division [String] Division name
    # @param level [Integer] Agent level (default: 4)
    # @param metadata [Hash] Additional metadata
    # @return [Hash] Created agent
    def register(name:, type: "ai", division: nil, level: 4, metadata: nil)
      body = { name: name, type: type, level: level }
      body[:division] = division if division
      body[:metadata] = metadata if metadata

      @client.post("/agents", body)
    end

    # Send a heartbeat for an agent
    #
    # @param agent_id [String] The agent ID
    # @param load [Float] Current load (0.0 to 1.0)
    # @return [Hash] Response
    def heartbeat(agent_id, load: nil)
      body = {}
      body[:load] = load if load
      @client.post("/agents/#{agent_id}/heartbeat", body)
    end

    # Update agent status
    #
    # @param agent_id [String] The agent ID
    # @param status [String] New status
    # @return [Hash] Updated agent
    def update_status(agent_id, status)
      @client.put("/agents/#{agent_id}", { status: status })
    end

    # Delete an agent
    #
    # @param agent_id [String] The agent ID
    # @return [Hash] Response
    def delete(agent_id)
      @client.delete("/agents/#{agent_id}")
    end

    # Get agent statistics
    #
    # @return [Hash] Statistics
    def stats
      @client.get("/agents/stats")
    end

    # Get agents by division
    #
    # @param division [String] Division name
    # @return [Array<Hash>] List of agents
    def by_division(division)
      list(division: division)
    end

    # Get Level 2 commanders
    #
    # @return [Array<Hash>] List of commander agents
    def commanders
      list(level: 2)
    end

    # Get Level 3 managers
    #
    # @return [Array<Hash>] List of manager agents
    def managers
      list(level: 3)
    end

    # Get Level 4 workers
    #
    # @return [Array<Hash>] List of worker agents
    def workers
      list(level: 4)
    end
  end
end

File: python/blackroad/agents.py
"""
BlackRoad Agents API
"""

from typing import Optional, Dict, Any, List


class AgentAPI:
    """
    Agent management API.

    Usage:
        # List all agents
        agents = client.agents.list()

        # Get specific agent
        agent = client.agents.get("agent-id")

        # Register a new agent
        agent = client.agents.register(
            name="my-agent",
            type="ai",
            division="OS"
        )

        # Send heartbeat
        client.agents.heartbeat("agent-id")

        # Update agent status
        client.agents.update_status("agent-id", "active")
    """

    def __init__(self, client):
        self._client = client

    def list(
        self,
        level: Optional[int] = None,
        division: Optional[str] = None,
        status: Optional[str] = None,
        limit: int = 100,
        offset: int = 0
    ) -> List[Dict[str, Any]]:
        """
        List agents with optional filters.

        Args:
            level: Filter by hierarchy level (1-4)
            division: Filter by division name
            status: Filter by status (active, standby, dead)
            limit: Maximum number of results
            offset: Pagination offset

        Returns:
            List of agent dictionaries
        """
        params = {"limit": limit, "offset": offset}
        if level:
            params["level"] = level
        if division:
            params["division"] = division
        if status:
            params["status"] = status

        return self._client.get("/agents", params=params).get("agents", [])

    def get(self, agent_id: str) -> Dict[str, Any]:
        """Get a specific agent by ID."""
        return self._client.get(f"/agents/{agent_id}")

    def register(
        self,
        name: str,
        agent_type: str = "ai",
        division: Optional[str] = None,
        level: int = 4,
        metadata: Optional[Dict] = None
    ) -> Dict[str, Any]:
        """
        Register a new agent.

        Args:
            name: Agent name
            agent_type: Type of agent (ai, hardware, human)
            division: Division assignment
            level: Hierarchy level (1-4)
            metadata: Additional metadata

        Returns:
            Created agent dictionary
        """
        data = {
            "name": name,
            "type": agent_type,
            "level": level
        }
        if division:
            data["division"] = division
        if metadata:
            data["metadata"] = metadata

        return self._client.post("/agents", data=data)

    def heartbeat(self, agent_id: str, load: Optional[float] = None) -> Dict[str, Any]:
        """
        Send a heartbeat for an agent.

        Args:
            agent_id: Agent ID
            load: Current load (0.0 to 1.0)

        Returns:
            Heartbeat response
        """
        data = {}
        if load is not None:
            data["load"] = load

        return self._client.post(f"/agents/{agent_id}/heartbeat", data=data)

    def update_status(self, agent_id: str, status: str) -> Dict[str, Any]:
        """
        Update agent status.

        Args:
            agent_id: Agent ID
            status: New status (active, standby, maintenance)

        Returns:
            Updated agent
        """
        return self._client.put(f"/agents/{agent_id}", data={"status": status})

    def delete(self, agent_id: str) -> Dict[str, Any]:
        """Delete an agent."""
        return self._client.delete(f"/agents/{agent_id}")

    def stats(self) -> Dict[str, Any]:
        """Get agent statistics."""
        return self._client.get("/agents/stats")

    def by_division(self, division: str) -> List[Dict[str, Any]]:
        """Get all agents in a division."""
        return self.list(division=division)

    def commanders(self) -> List[Dict[str, Any]]:
        """Get all Level 2 commanders."""
        return self.list(level=2)

    def managers(self) -> List[Dict[str, Any]]:
        """Get all Level 3 managers."""
        return self.list(level=3)

    def workers(self) -> List[Dict[str, Any]]:
        """Get all Level 4 workers."""
        return self.list(level=4)
